File: oss/clawpipe/go-sdk/router.go
package clawpipe

import (
	"math"
	"regexp"
	"sort"
	"sync"
)

// ModelProfile describes a provider/model with cost and quality data.
type ModelProfile struct {
	Provider        string
	Model           string
	CostPer1kTokens float64
	AvgLatencyMs    float64
	QualityScore    float64
	MaxTokens       int
}

// learnedWeight tracks per-model outcomes.
type learnedWeight struct {
	TotalCalls  int
	AvgLatency  float64
	AvgTokenOut float64
	Score       float64
}

type taskComplexity int

const (
	simple  taskComplexity = iota
	medium
	complex
)

// DefaultModels is the built-in model catalogue.
var DefaultModels = []ModelProfile{
	{"deepseek", "deepseek-chat", 0.14, 800, 0.82, 64000},
	{"openai", "gpt-4o-mini", 0.15, 600, 0.85, 128000},
	{"anthropic", "claude-3-haiku", 0.25, 500, 0.88, 200000},
	{"openai", "gpt-4o", 2.5, 1200, 0.94, 128000},
	{"anthropic", "claude-sonnet-4", 3.0, 1000, 0.95, 200000},
	{"anthropic", "claude-opus-4", 15.0, 2000, 0.99, 200000},
	{"groq", "llama-3.1-70b", 0.59, 300, 0.80, 32000},
	{"mistral", "mistral-large", 2.0, 900, 0.90, 128000},
}

// Router selects the best provider/model for a prompt.
type Router struct {
	mu      sync.RWMutex
	models  []ModelProfile
	weights map[string]*learnedWeight
}

// NewRouter creates a Router with the default model catalogue.
func NewRouter() *Router {
	return &Router{models: DefaultModels, weights: map[string]*learnedWeight{}}
}

// Route picks the best model for the prompt and options.
func (r *Router) Route(prompt string, opts *PromptOptions) RouteDecision {
	if opts != nil && opts.Model != "" && opts.Provider != "" {
		return RouteDecision{opts.Provider, opts.Model, 1, "explicit"}
	}
	cx := classifyComplexity(prompt)
	candidates := r.rankCandidates(cx)

	r.mu.RLock()
	scored := make([]scoredModel, len(candidates))
	for i, c := range candidates {
		scored[i] = c
		key := c.Provider + ":" + c.Model
		if lw, ok := r.weights[key]; ok {
			scored[i].score += (lw.Score - 0.5) * 0.2
		}
	}
	r.mu.RUnlock()

	sort.Slice(scored, func(i, j int) bool { return scored[i].score > scored[j].score })
	b := scored[0]
	tag := "simple"
	if cx == medium {
		tag = "medium"
	} else if cx == complex {
		tag = "complex"
	}
	return RouteDecision{b.Provider, b.Model, b.score, "complexity=" + tag}
}

// Learn records an outcome for self-learning routing.
func (r *Router) Learn(rd RouteDecision, latencyMs float64, tokensOut int) {
	r.mu.Lock()
	defer r.mu.Unlock()
	key := rd.Provider + ":" + rd.Model
	ex, ok := r.weights[key]
	if !ok {
		r.weights[key] = &learnedWeight{
			TotalCalls: 1, AvgLatency: latencyMs,
			AvgTokenOut: float64(tokensOut),
			Score:       computeLearnScore(latencyMs, float64(tokensOut)),
		}
		return
	}
	n := float64(ex.TotalCalls + 1)
	ex.AvgLatency += (latencyMs - ex.AvgLatency) / n
	ex.AvgTokenOut += (float64(tokensOut) - ex.AvgTokenOut) / n
	ex.TotalCalls = int(n)
	ex.Score = computeLearnScore(ex.AvgLatency, ex.AvgTokenOut)
}

var (
	codeRe = regexp.MustCompile("```[\\s\\S]+```|(?:function|class|const)\\s")
	stepRe = regexp.MustCompile(`(?i)\b(then|after that|next|finally|step \d)\b`)
)

func classifyComplexity(prompt string) taskComplexity {
	tokens := (len(prompt) + 3) / 4
	hasCode := codeRe.MatchString(prompt)
	hasMulti := stepRe.MatchString(prompt)
	if tokens > 2000 || (hasCode && hasMulti) {
		return complex
	}
	if tokens > 500 || hasCode || hasMulti {
		return medium
	}
	return simple
}

type scoredModel struct {
	ModelProfile
	score float64
}

func (r *Router) rankCandidates(cx taskComplexity) []scoredModel {
	var costW, qualW, speedW float64
	switch cx {
	case simple:
		costW, qualW = 0.6, 0.2
	case medium:
		costW, qualW = 0.3, 0.5
	default:
		costW, qualW = 0.1, 0.7
	}
	speedW = 1 - costW - qualW

	out := make([]scoredModel, len(r.models))
	for i, m := range r.models {
		cs := 1 - math.Min(m.CostPer1kTokens/15, 1)
		qs := m.QualityScore
		ss := 1 - math.Min(m.AvgLatencyMs/3000, 1)
		out[i] = scoredModel{m, costW*cs + qualW*qs + speedW*ss}
	}
	return out
}

func computeLearnScore(latMs, tokOut float64) float64 {
	ls := 1 - math.Min(latMs/5000, 1)
	es := math.Min(tokOut/1000, 1)
	return ls*0.5 + es*0.5
}

// ModelCount returns the number of known models.
func (r *Router) ModelCount() int { return len(r.models) }

// Complexity exports the classifier for testing.
func Complexity(prompt string) string {
	c := classifyComplexity(prompt)
	return []string{"simple", "medium", "complex"}[c]
}

File: oss/clawpipe/sdk/src/router.ts
/**
 * Smart Router — self-learning model selection with optional global weight sync.
 *
 * Tracks outcomes per task type and routes prompts to the best provider/model
 * based on cost, quality, and latency. Weights update after every call and can
 * optionally sync to the ClawPipe gateway for cross-instance learning.
 */

import { fetchRemoteWeights, pushRemoteWeights, pushQualityScore } from './router-sync';
import { healthPenalty, type HealthMap, DEFAULT_FALLBACK_COUNT } from './failover';
import { DEFAULT_MODELS, type ModelProfile } from './router-models';

export interface RouterConfig {
  models?: ModelProfile[];
  gatewayUrl?: string;
  apiKey?: string;
  /** When true, push/fetch learned weights from the gateway. Default: false (opt-in). */
  globalLearning?: boolean;
  /** Fraction of responses to quality-score via LLM-as-judge. Default: 0.1 */
  scoringSampleRate?: number;
}

export interface RouteDecision {
  provider: string;
  model: string;
  score: number;
  reason: string;
}

export interface LearnedWeight {
  totalCalls: number;
  avgLatencyMs: number;
  avgTokensOut: number;
  score: number;
}

type TaskComplexity = 'simple' | 'medium' | 'complex';

/** Merge two weight maps using weighted average by totalCalls. */
export function mergeWeights(
  local: Record<string, LearnedWeight>,
  remote: Record<string, LearnedWeight>,
): Record<string, LearnedWeight> {
  const keys = new Set([...Object.keys(local), ...Object.keys(remote)]);
  const result: Record<string, LearnedWeight> = {};
  for (const key of keys) {
    const l = local[key];
    const r = remote[key];
    if (!l) { result[key] = r; continue; }
    if (!r) { result[key] = l; continue; }
    const total = l.totalCalls + r.totalCalls;
    result[key] = {
      totalCalls: total,
      score:        (l.score        * l.totalCalls + r.score        * r.totalCalls) / total,
      avgLatencyMs: (l.avgLatencyMs * l.totalCalls + r.avgLatencyMs * r.totalCalls) / total,
      avgTokensOut: (l.avgTokensOut * l.totalCalls + r.avgTokensOut * r.totalCalls) / total,
    };
  }
  return result;
}

export class Router {
  private models: ModelProfile[];
  private weights = new Map<string, LearnedWeight>();
  private config: RouterConfig;
  private weightsLoaded = false;
  /** Per-Router health map — mutated by failover.runWithFailover() callers. */
  readonly health: HealthMap = new Map();

  constructor(config?: RouterConfig) {
    this.config = config ?? {};
    this.models = config?.models ?? DEFAULT_MODELS;
  }

  /** Route a prompt to the best provider/model. Fetches remote weights on first call if globalLearning is on. */
  async route(
    prompt: string,
    options: { model?: string; provider?: string; taskType?: string } = {},
  ): Promise<RouteDecision> {
    await this.loadRemoteWeights();
    if (options.model && options.provider) {
      return { provider: options.provider, model: options.model, score: 1, reason: 'explicit' };
    }
    const ranked = this.rankWithComplexity(prompt);
    const best = ranked[0];
    return { provider: best.provider, model: best.model, score: best.score, reason: best.reason };
  }

  /** Up to `count` next-best routes, excluding `primary`. Used for failover. */
  fallbacks(primary: RouteDecision, prompt: string, count: number = DEFAULT_FALLBACK_COUNT): RouteDecision[] {
    const ranked = this.rankWithComplexity(prompt);
    const out: RouteDecision[] = [];
    for (const r of ranked) {
      if (r.provider === primary.provider && r.model === primary.model) continue;
      out.push({ provider: r.provider, model: r.model, score: r.score, reason: r.reason });
      if (out.length >= count) break;
    }
    return out;
  }

  /** Internal: rank all models for a prompt with complexity, learned bonus, and health penalty. */
  private rankWithComplexity(prompt: string): Array<ModelProfile & { score: number; reason: string }> {
    const complexity = this.classifyComplexity(prompt);
    const candidates = this.rankCandidates(complexity);
    const scored = candidates.map((c) => {
      const key = `${c.provider}:${c.model}`;
      const learned = this.weights.get(key);
      const learnedBonus = learned ? (learned.score - 0.5) * 0.2 : 0;
      const penalty = healthPenalty(this.health, c.provider);
      return { ...c, score: c.score + learnedBonus - penalty, reason: `complexity=${complexity}` };
    });
    return scored.sort((a, b) => b.score - a.score);
  }

  /** Record outcome. Optionally push updated weights to gateway (fire-and-forget). */
  learn(route: RouteDecision, latencyMs: number, tokensOut: number, qualityScore?: number): void {
    const key = `${route.provider}:${route.model}`;
    const existing = this.weights.get(key);
    if (!existing) {
      this.weights.set(key, {
        totalCalls: 1, avgLatencyMs: latencyMs, avgTokensOut: tokensOut,
        score: this.computeScore(latencyMs, tokensOut, qualityScore),
      });
    } else {
      const n = existing.totalCalls + 1;
      existing.avgLatencyMs = existing.avgLatencyMs + (latencyMs - existing.avgLatencyMs) / n;
      existing.avgTokensOut = existing.avgTokensOut + (tokensOut - existing.avgTokensOut) / n;
      existing.totalCalls = n;
      existing.score = this.computeScore(existing.avgLatencyMs, existing.avgTokensOut, qualityScore);
      this.weights.set(key, existing);
    }
    pushRemoteWeights(this.config, this.getAllWeights()).catch(() => {});
  }

  /** Push a quality score to the gateway. Fire-and-forget — never throws. */
  pushQualityScore(payload: { request_id: string; model: string; provider: string; score: number }): void {
    pushQualityScore(this.config, payload);
  }

  getWeights(): Map<string, LearnedWeight> { return new Map(this.weights); }
  setWeights(weights: Map<string, LearnedWeight>): void { this.weights = new Map(weights); }

  /** Get all weights as a plain object (for gateway push). */
  getAllWeights(): Record<string, LearnedWeight> {
    const out: Record<string, LearnedWeight> = {};
    for (const [k, v] of this.weights) out[k] = v;
    return out;
  }

  // ── Private ────────────────────────────────────────────────────────────────

  /** Lazy one-time fetch of remote weights. No-op if globalLearning is false. */
  private async loadRemoteWeights(): Promise<void> {
    if (!this.config.globalLearning || !this.config.gatewayUrl || this.weightsLoaded) return;
    this.weightsLoaded = true;
    const merged = await fetchRemoteWeights(this.config, this.getAllWeights());
    if (merged) this.weights = new Map(Object.entries(merged));
  }

  private computeScore(latencyMs: number, tokensOut: number, qualityScore?: number): number {
    const latencyScore = 1 - Math.min(latencyMs / 5000, 1);
    const efficiencyScore = Math.min(tokensOut / 1000, 1);
    if (qualityScore !== undefined) {
      return latencyScore * 0.4 + efficiencyScore * 0.3 + qualityScore * 0.3;
    }
    return latencyScore * 0.5 + efficiencyScore * 0.5;
  }

  private classifyComplexity(prompt: string): TaskComplexity {
    const tokens = Math.ceil(prompt.length / 4);
    const hasCode = /```[\s\S]+```/.test(prompt) || /function\s|class\s|const\s/.test(prompt);
    const hasMultiStep = /\b(then|after that|next|finally|step \d)\b/i.test(prompt);
    if (tokens > 2000 || (hasCode && hasMultiStep)) return 'complex';
    if (tokens > 500 || hasCode || hasMultiStep) return 'medium';
    return 'simple';
  }

  private rankCandidates(complexity: TaskComplexity): (ModelProfile & { score: number })[] {
    const costWeight    = complexity === 'simple' ? 0.6 : complexity === 'medium' ? 0.3 : 0.1;
    const qualityWeight = complexity === 'simple' ? 0.2 : complexity === 'medium' ? 0.5 : 0.7;
    const speedWeight   = 1 - costWeight - qualityWeight;
    return this.models.map((m) => {
      const costScore    = 1 - Math.min(m.costPer1kTokens / 15, 1);
      const qualityScore = m.qualityScore;
      const speedScore   = 1 - Math.min(m.avgLatencyMs / 3000, 1);
      return { ...m, score: costWeight * costScore + qualityWeight * qualityScore + speedWeight * speedScore };
    });
  }
}
